### PR TITLE
Add project model and ddb functions

### LINKED ISSUE #185: Create data model for projects
The model should be created as per this diagram here.
image

### PR DESCRIPTION
**Pull request by cheukyin699**
# Pull Request

## Progress

- [x] Project model
- [x] Project model tests
- [x] DynamoDB functions that operate on Project model
- [x] DynamoDB functions tests

## Description

Add the project model and a few helper functions, along with a bunch of tests. Also add some DynamoDB functions that operate on the project model.

## Ticket(s)

Closes #185.

### CODE REVIEW

File: config.toml [hunk 1]
```diff
@@ -7,8 +7,9 @@ creds_path = "credentials/slack.toml"
 creds_path = "credentials/aws.toml"
 users_table = "users"
 teams_table = "teams"
+teams_table = "projects"
```
Comment (rwblickhan) [inline config.toml]: This should be `projects_table`, right?

File: db/dynamodb.py [hunk 2]
```diff
@@ -104,6 +108,38 @@ def __create_user_tables(self):
...
+
+        **Note**: This function should **not** be called externally, and should
+        only be called on initialization, if at all.
+
+        Projects are only required to have a ``project_id``, technically. But we
+        also require that there be at least 1 URL in the ``github_urls`` field.
+        This is not programmically enforced.
+        """
+        logging.info("Creating table '{}'".format(self.projects_table))
+        self.ddb.create_table(
```
Comment (rwblickhan) [inline db/dynamodb.py]: As a side note, we might consider refactoring these into [CloudFormation](https://aws.amazon.com/cloudformation/) files - anything we request can be a simple JSON config file (that looks an awful lot like this request anyway) and then have a single "resource request" function that just reads the CloudFormation files and makes the request through CF.

**Comment by cheukyin699** [comment]
There is a lot of duplicated code in the `db/dynamodb.py` file. The code structures of `store_`, `retrieve_`, `query_`, and `delete_` are almost the same, and I think we can abstract most of the cruft out so that the file is no longer 400 lines long. Might add that as an issue.

File: docs/Database.md [hunk 3]
```diff
@@ -34,3 +34,23 @@ Attribute Name | Description
...
+are also listed here:
+
+Attribute Name | Description
+---|---
+`project_id` | `String`; The project's unique SHA1 ID, salted with a timestamp
+`github_team_id` | `String`; The team's Github ID associated with the project
+`github_urls` | `String Set`; A set of URLs pointing to project repositories
+`display_name` | `String`; A name for the project
+`short_description` | `String`; A short description that outlines the project
+`long_desscription` | `String`; A longer and more in-depth description
```
Comment (jacqueschen1) [inline docs/Database.md]: Nit: typo :p

File: model/project.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+class Project:
+    """Represent a team project with team ID and related fields and methods."""
+
+    def __init__(self, github_team_id, github_urls):
+        """
+        Initialize the team project.
+
+        Project ID is just the current epoch time and the first github project
+        URL mashed together using SHA1.
```
Comment (omar2535) [inline model/project.py]: Can't you just say `project id is sha1 generated from github project url and epoch time`?

Comment (cheukyin699) [reply model/project.py]: ok

File: docs/Database.md [hunk 5]
```diff
@@ -34,3 +34,23 @@ Attribute Name | Description
...
+
+The `projects` table stores all projects where `project_id` is the primary
+index. All other attributes are specified in the `model/project.py` file, and
+are also listed here:
+
+Attribute Name | Description
+---|---
+`project_id` | `String`; The project's unique SHA1 ID, salted with a timestamp
+`github_team_id` | `String`; The team's Github ID associated with the project
+`github_urls` | `String Set`; A set of URLs pointing to project repositories
```
Comment (jacqueschen1) [inline docs/Database.md]: Note for future: it may be best to store the github repo IDs instead of urls

File: model/project.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+        Project ID is a SHA1 generated from the first Github project URL and
+        epoch time.
+
+        :param github_team_id: the Github team ID associated with the project
+        :param github_urls: a set/list of URLs pointing to repositories
+        """
+        hfunc = hashlib.sha1()
+        hfunc.update(bytes(github_urls[0], 'utf-8'))
+        hfunc.update(bytes(str(time.time()), 'utf-8'))
```
Comment (rwblickhan) [inline model/project.py]: Out of curiosity, is there a reason you did this instead of, say, a UUID?

**Pull request merged**